Integrations tests and helper functions

itirabasso opened the pull request:
Merge after #43

plaintextpaco commented on src/test/User.sol:
```diff
@@ -0,0 +1,30 @@
+import {Utils} from "./Utils.sol";
+
+contract User {
+    // string memory mnemonic = "test test test test test test test test test test test junk";
```
remove commentend code

plaintextpaco commented on src/test/Utils.sol:
```diff
@@ -1,67 +1,210 @@
+        );
+    }
+
+    function createInstance(address owner, CrucibleFactory crucibleFactory) internal returns (Crucible crucible) {
```
this function is unused

plaintextpaco commented on src/test/Utils.sol:
```diff
@@ -1,67 +1,210 @@
+        return Vm(address(0x7109709ECfa91a80626fF3989D68f67F5b1DD12D));
+    }
+
+    function launchProgram(
```
this function is unused

itirabasso replied:
This is used in #40

plaintextpaco commented on src/test/Utils.sol:
```diff
@@ -1,67 +1,210 @@
+        aludel.unstakeAndClaim(address(crucible), amount, unlockSig);
+    }
+
+    function fundMockToken(
```
this function is unused

itirabasso replied:
It's only for the sake of sanity.

Make some lines short because removes the need to cast the ERC20 token to address.
It's used in #40 tests

plaintextpaco commented on src/test/AludelFactoryIntegration.sol:
```diff
@@ -61,17 +64,24 @@ contract AludelFactoryIntegrationTest is DSTest {
         bps = 100;
-        factory = new AludelFactory(recipient, bps);
+
+        UserManager userManager = new UserManager();
```
any reason for the UserManager not to store the current index and make it auto-incremental? would it ever be used in a different way?

itirabasso replied:
I left that for later

plaintextpaco commented on src/test/AludelFactoryIntegration.sol:
```diff
@@ -61,17 +64,24 @@ contract AludelFactoryIntegrationTest is DSTest {
-        owner = cheats.addr(PRIVATE_KEY);
-
-        recipient = cheats.addr(PRIVATE_KEY + 1);
+        vm = Vm(HEVM_ADDRESS);
```
Given that the `Utils` file is imported here I think we could do without the `HEVM_ADDRESS` and just call `Utils.vm()`

itirabasso replied:
Yeees

plaintextpaco commented on src/test/AludelFactoryIntegration.sol:
```diff
@@ -115,269 +125,111 @@ contract AludelFactoryIntegrationTest is DSTest {
-        aludel.fund(1 ether, 1 days);
-        aludel.registerVaultFactory(address(template));
-        cheats.stopPrank();
+        // todo : add fund aludel helper function
```
this wasn't implemented. Since it's a small thing I'd advise to do it here, but feel free to create an issue to tackle it later

itirabasso replied:
Implemented.

plaintextpaco commented on src/test/AludelFactoryIntegration.sol:
```diff
@@ -115,269 +125,111 @@ contract AludelFactoryIntegrationTest is DSTest {
 
-        vault = aludel.getVaultData(crucible);
-        assertEq(vault.totalStake, 1 ether);
+        vault = aludel.getVaultData(address(crucible));
```
I think all of the assertions that are a consequence of _staking_ should go in the previous testcase, and this one should assume staking works okay and focus on asserting what happens _after_ unstaking

plaintextpaco reviewed (changes requested):
commits
-------
- [x] bc48358 update ds-test and forge-std -- only updates submodules
- [x] bd94889 update solmate
    - [ ] I think this commit and the previous one could've been one, they don't really do a lot and even if the changes in the rest of the codebase are included here as suggeted, it would still be easy to follow and the changes from updating every dependency are probably kinda related anyways
- [x] 784b0d3 add user manager and user contracts
    - [ ] commented code uwu
    - [ ] is it a user _manager_ or a user _factory_? because the only thing that it does is to create Users with different indices. I'll have to see how it's used to figure out if i like it
- [x] 43b1cc1 add helper functions for aludel tests
- [x] a661353 use helper functions in integration tests
    - [ ] not really asking you fix it now, but please do implmenent it in the future: this commit includes changes which are actually caused by updating solmate, ds-test and forge-std. Those should've been included in their respective commits
    - [ ] fundAludel helper function: is it impemented eventually? wasn't it added in previous commit?
    - [ ] I don't get if the testcases were removed because they were redundant or hard to reimplement with the new abstractions?
- [x] 038abcc remove PRIVATE_KEY constant
    - [ ] it's part of the refactor of using the User and UserManager, doesn't really merit it's own commit
- [ ] db8ec19 replace hevm with vm
    - [ ] it'd have been cooler if you did all of the s/hevm/vm/ in a single commit. There is some of this which was done on a661353 use helper functions in integration tests, and having it taken care of in a different commit would've made that one slightly more readable
- [ ] 09a6b0a rename tests
    - [ ] there are some changes (eg, s/1 ether/STAKE_AMOUNT) which really should've been part of a previous commit, 'a661353 use helper functions in integration tests'
    - [ ] the coolest way to do the test deletions I think would've been to do all of them on a single commit and have the commit message explain _why_ it was no longer necessary.
- [ ] d30faec use double quotes for literal strings
    - [ ] there were plenty of cosmetic changes in '43b1cc1 add helper functions', I think the best would've been to first do all the cosmetic changes to the file in a single commit and then add functionality

comments
--------
- [ ] several of the functions defined in Utils.sol are unused. This is another argument in favor of creating and using every functionaly in a separate commit, instead of having one for creating them all and another one to use them.
- [ ] I really liked that we finally modularized the permission signing away from a particular test file.
- [ ] if I comment the locking of Crucible funds, then the staking test passes but the unstaking one fails with a somewhat-related error due to the funds not being unlocked in the first place
- [ ] if I comment the unlocking of Crucible funds, then all tests fail

Blocking issues
---------------
- [ ] move the asserts for staking into the first test
- [ ] rebase on top of main
- [ ] remove dead code
- [ ] assert Crucible funds are locked and unlocked

plaintextpaco commented on src/test/Utils.sol:
```diff
@@ -1,67 +1,210 @@
+        return Crucible(payable(crucibleFactory.create("")));
+    }
+
+    function createInstance(User user, CrucibleFactory crucibleFactory) internal returns (Crucible crucible) {
```
It's not obvious to me an instance _of what_ it's creating. I'd advise to rename it to createCrucible or something like that

itirabasso replied:
This comes from the old alchemist hardhat tasks, `create-instance` is the task used to create crucible instance. I've updated the name to `createCrucible`.

itirabasso commented:
The unused util functions are actually used in the Aludel V2 solidity tests, which are part of a upcoming PR but because AludelFactoryIntegration.sol uses _some_ of them I included the whole file in this PR. In my opinion, splitting the file would involved undesired rework.

I think checking the Aludel locks and unlocks crucible funds is more suitable for the Aludel test cases.

itirabasso commented:
I'm not entire sure about using gherkin for solidity tests. It's starting to be more cumbersome than useful. I agree that we need to define better, clearer scenarios/contexts in which the test take place but the reusability proposed by gherkin it's not easy to achieve in solidity.

plaintextpaco commented on src/test/AludelFactory.t.sol:
```diff
@@ -17,7 +17,7 @@ import "forge-std/src/console2.sol";
 contract AludelFactoryTest is DSTest {
     AludelFactory private factory;
-    Hevm private cheats;
+    Vm private cheats;
```
If this contract extends Test instead of DSTest, then this is automatically configured

plaintextpaco reviewed (changes requested):
> I think checking the Aludel locks and unlocks crucible funds is more suitable for the Aludel test cases than in AludelFactory tests

strongly agree for unitary tests, less strongly (but still) agree for integration tests

> I'm not entire sure about using gherkin for solidity tests. It's starting to be more cumbersome than useful. I agree that we need to define better, clearer scenarios/contexts in which the test take place but the reusability proposed by gherkin it's not easy to achieve in solidity.

The main purpose of gherkin is _clarity_, so by glancing at the tests' description you can get a good idea of _what_ they do at a domain level.

The reusability is an awesome side effect when paired with nestable testcases, but we can't achieve that with foundry.

I'm open to having test descriptions that don't look like gherkin, of course, as long as they provide some info about _what_ is being tested. I'm somewhat confident that they'll look similar though, since the _what_ is defined by declaring the context (GIVEN), an action (WHEN) and its consequences (THEN). What I think makes the most sense is:

- further divide the tests, having several unit test contracts per implementation contract instead of just one. These test contracts would of course have their own, different `setUp`s, in which they would get a few different GIVENs out of the way, which is what currently pollutes our test names the most.
- the testcases themselves could perhaps have several assertions, but they would test one _case_ of a particular _feature_.

The integration tests, while much cleaner now, don't actually assert anything beyond the transaction not reverting, I can for example remove the entirety of the reward-paying logic and the tests still pass. Having some assertions regarding system state after performing the test's actions is the only blocking issue.

itirabasso commented:
Rebased on top of #43. Moved out the dependencies/libs related commits to #43 to separate things a little bit more.

I added three commits:
`a745c4`: util function for aludel v3 unstaking
`e4a284`: adapt previous tests (made for aludel v2) to work on the current version v3
`fdd6f7`: imports clean up
